### PR TITLE
fix: Select the latest ModelRollout in the PDP instead of the oldest

### PR DESCRIPTION
**Pull request by ugur47**
Issue: #4448

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt [hunk 1]
```diff
@@ -211,18 +214,12 @@ class PopulationRequisitionFulfiller(
...
-          if (b.hasGradualRolloutPeriod()) b.gradualRolloutPeriod.endDate else b.instantRolloutDate
-        if (aDate.toLocalDate().isBefore(bDate.toLocalDate())) -1 else 1
-      }
-
-    // Retrieves latest ModelRollout from list.
-    val latestModelRollout = sortedModelRolloutsList.first()
+    // ModelRollouts for a ModelLine commonly share a rollout date, so createTime breaks ties in
+    // favor of the most recently created one.
+    val latestModelRollout: ModelRollout =
+      listModelRolloutsResponse.modelRolloutsList.maxWithOrNull(
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt]: Use the existing `modelRolloutsStub.listResources { ... }.flattenConcat()` pagination pattern here before applying the comparator. `ListModelRollouts` returns at most 50 resources by default and orders them by `rollout_start_time` ascending, so this currently selects the newest rollout from the oldest page rather than the actual latest rollout.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt [hunk 2]
```diff
@@ -376,6 +381,169 @@ class PopulationRequisitionFulfillerTest {
     assertThat(result.population.value).isEqualTo(VID_RANGE_3.size())
   }
 
+  @Test
+  fun `uses ModelRollout with latest rollout date`() {
```
Comment (stevenwarejones) [inline src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt:385]: Add a regression test that stubs a first response with a non-empty `next_page_token` and places the latest rollout in the second response. The current tests only return one page, so they do not catch this bug.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt [hunk 3]
```diff
@@ -268,3 +265,9 @@ class PopulationRequisitionFulfiller(
     fulfillDirectMeasurement(requisition, measurementSpec, requisitionSpec.nonce, measurementResult)
   }
 }
+
+/** Date on which this [ModelRollout] has taken full effect. */
+private val ModelRollout.rolloutDate: LocalDate
+  get() =
+    if (hasGradualRolloutPeriod()) gradualRolloutPeriod.endDate.toLocalDate()
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt]: Order gradual rollouts by `gradualRolloutPeriod.startDate`, not `endDate`. `ListModelRollouts` and Kingdom's `previous_model_rollout` relationship define rollout order by `rollout_start_time`. Using the end date lets an older long-running gradual rollout outrank a newer rollout that starts before it finishes, selecting a superseded `ModelRelease`. Rename this property to `rolloutStartDate`.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt [hunk 4]
```diff
@@ -376,6 +381,169 @@ class PopulationRequisitionFulfillerTest {
...
+
+    runBlocking { requisitionFulfiller.executeRequisitionFulfillingWorkflow() }
+
+    val request: GetModelReleaseRequest =
+      verifyAndCapture(modelReleasesServiceMock, ModelReleasesCoroutineImplBase::getModelRelease)
+    assertThat(request.name).isEqualTo(MODEL_RELEASE_NAME_2)
+  }
+
+  @Test
+  fun `uses end of gradual rollout period as rollout date`() {
```
Comment (stevenwarejones) [inline src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt]: Replace this with a regression test where an older gradual rollout starts first but ends after a newer instant rollout starts, and assert that the newer instant rollout wins. The current test codifies the incorrect end-date ordering.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt [hunk 5]
```diff
@@ -42,6 +44,7 @@ import org.wfanet.measurement.api.v2alpha.getModelReleaseRequest
 import org.wfanet.measurement.api.v2alpha.getPopulationRequest
 import org.wfanet.measurement.api.v2alpha.listModelRolloutsRequest
 import org.wfanet.measurement.common.throttler.Throttler
+import org.wfanet.measurement.common.toInstant
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt:54]: Add `@wfa_common_jvm//src/main/kotlin/org/wfanet/measurement/common` to the `population_requisition_fulfiller` target's `deps`. This file directly imports `toInstant` and `toLocalDate`; relying on another dependency to expose that library transitively violates the explicit-dependency rule in `docs/bazel-build-standards.md`.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt [hunk 6]
```diff
@@ -100,6 +104,7 @@ import org.wfanet.measurement.common.pack
 import org.wfanet.measurement.common.readByteString
 import org.wfanet.measurement.common.testing.verifyAndCapture
 import org.wfanet.measurement.common.throttler.Throttler
+import org.wfanet.measurement.common.toProtoDate
```
Comment (stevenwarejones) [inline src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt:107]: Add `@wfa_common_jvm//src/main/kotlin/org/wfanet/measurement/common` to the `population_requisition_fulfiller_test` target's `deps`. This test directly imports `toProtoDate` and other helpers from that library, so it must declare the dependency explicitly.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt (hunk 5 above)
Comment (ugur47) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt:54]: Done.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt (hunk 1 above)
Comment (ugur47) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt]: Done.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt (hunk 3 above)
Comment (ugur47) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt]: Done.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt (hunk 6 above)
Comment (ugur47) [reply src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt:107]: Done.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt (hunk 2 above)
Comment (ugur47) [reply src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt:385]: Done.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt (hunk 4 above)
Comment (ugur47) [reply src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt]: Done.

**Review by ugur47 — COMMENTED** [review]
> @ugur47 made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on SanjayVas and stevenwarejones).

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel [hunk 7]
```diff
@@ -17,6 +17,7 @@ kt_jvm_library(
     srcs = ["PopulationRequisitionFulfiller.kt"],
     deps = [
         "//src/main/kotlin/org/wfanet/measurement/api/v2alpha:population_spec_validator",
+        "//src/main/kotlin/org/wfanet/measurement/common/api/grpc:list_resources",
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel:20]: Add `@wfa_common_jvm//imports/kotlin/kotlinx/coroutines:core` to this target's `deps`. `PopulationRequisitionFulfiller.kt` now directly imports `ExperimentalCoroutinesApi`, `Flow`, and `toList`; currently those symbols are available only transitively through `:list_resources`, which violates `docs/bazel-build-standards.md`.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt (hunk 5 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt:54]: Verified - the production target now directly declares the common-jvm dependency used by `toInstant` and `toLocalDate`.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt (hunk 1 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt]: Verified - `getModelRelease` now follows every `ListModelRollouts` page with `listResources`/`flattenConcat` before selecting the maximum.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt (hunk 3 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt]: Verified - gradual rollouts now compare `gradualRolloutPeriod.startDate`, matching the Kingdom ordering.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt (hunk 6 above)
Comment (stevenwarejones) [reply src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt:107]: Verified - the test target now directly declares both the Date proto and common-jvm dependencies used by the new test code.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt (hunk 2 above)
Comment (stevenwarejones) [reply src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt:385]: Verified - the regression test puts the latest rollout on the second page behind a non-empty `next_page_token`.

File: src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt (hunk 4 above)
Comment (stevenwarejones) [reply src/test/kotlin/org/wfanet/measurement/loadtest/dataprovider/PopulationRequisitionFulfillerTest.kt]: Verified - the overlap regression test now confirms a newer instant rollout wins even when the earlier gradual rollout ends later.

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> @stevenwarejones+AGNT made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on SanjayVas, stevenwarejones, and ugur47).

**Review by stevenwarejones — COMMENTED** [review]
> @stevenwarejones reviewed 4 files and all commit messages, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: all files reviewed, 6 unresolved discussions (waiting on SanjayVas and ugur47).

**Review by stevenwarejones — COMMENTED** [review]
> @stevenwarejones resolved 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: all files reviewed, 1 unresolved discussion (waiting on SanjayVas and ugur47).

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel (hunk 7 above)
Comment (ugur47) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel:20]: Done.

**Review by ugur47 — COMMENTED** [review]
> @ugur47 made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on SanjayVas, stevenwarejones, and stevenwarejones+AGNT).

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> @SanjayVas reviewed all commit messages and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on stevenwarejones, stevenwarejones+AGNT, and ugur47).

___
*[`src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt` line 241 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449#-P0Y-bbD9E2gnfVL_sYR:-P0Y-bbD9E2gnfVL_sYS:bphksse) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/8d28da1fe53eca91295d276d846c2154c9fb3f86/src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt#L241)):*
> ```Kotlin
>       modelRolloutLists
>         .flattenConcat()
>         .toList()
> ```

This will hold all the responses in memory at once. You need not do this to compute a maximum. You can operate on the flow without converting to a list.

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel (hunk 7 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel:20]: Verified - the production target now directly declares the coroutines dependency used by `ExperimentalCoroutinesApi`, `Flow`, and `toList`.

**Review by stevenwarejones — COMMENTED** [review]
> @stevenwarejones+AGNT made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on stevenwarejones and ugur47).

**Review by stevenwarejones — APPROVED** [review]

File: src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel (hunk 7 above)
Comment (ugur47) [reply src/main/kotlin/org/wfanet/measurement/populationdataprovider/BUILD.bazel:20]: Done.

**Review by ugur47 — COMMENTED** [review]
> @ugur47 made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on SanjayVas and stevenwarejones).

___
*[`src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt` line 241 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449#-P0Y-bbD9E2gnfVL_sYR:-P0Y2j46BCjbwCYYtKfJ:b-rnkj8a) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/8d28da1fe53eca91295d276d846c2154c9fb3f86/src/main/kotlin/org/wfanet/measurement/populationdataprovider/PopulationRequisitionFulfiller.kt#L241)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

This will hold all the responses in memory at once. You need not do this to compute a maximum. You can operate on the flow without converting to a list.
</blockquote></details>

Done — folded over the flow with a running max instead, so nothing is materialized.

**Review by ugur47 — COMMENTED** [review]
> @ugur47 partially reviewed 4 files.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on SanjayVas and stevenwarejones).

**Review by SanjayVas — APPROVED** [review]
> @SanjayVas reviewed 4 files and all commit messages, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: all files reviewed, 1 unresolved discussion (waiting on stevenwarejones).

**Review by ugur47 — COMMENTED** [review]
> @ugur47 reviewed 1 file and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449)* status: all files reviewed, 1 unresolved discussion (waiting on stevenwarejones).

**Review by ugur47 — COMMENTED** [review]
:lgtm:

> @ugur47 made 1 comment, resolved 1 discussion, and dismissed @stevenwarejones and @stevenwarejones+AGNT from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4449#-:-P0Y6-gZ6TVFIO059px5:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on stevenwarejones).
